### ISSUE TITLE
BUG: special.chdtrc: returns 1.0 when both degrees of freedom and argument are negative

### ISSUE THREAD

**Issue by mdhaber** [scipy/scipy#20972]
### Describe your issue.

As expected, the CDF of the chi-squared distribution is undefined for negative degrees of freedom.
```python3
from scipy import special
special.chdtr(-1, [-1, 1])  # array([nan, nan])
```

Yet:
```python3
special.chdtrc(-1, [-1, 1])  # array([1., nan])
```

### Reproducing Code Example

```python
from scipy import special
special.chdtrc(-1, -1)  # produces 1.0 instead of NaN
```

### Error message

```shell
NA
```

### SciPy/NumPy/Python version and system information

```shell
1.15.0.dev0+935.e984c95 1.26.4 sys.version_info(major=3, minor=12, micro=3, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: C:/Users/matth/miniforge3/envs/scipy-dev/Library/include/openblas
    lib directory: C:/Users/matth/miniforge3/envs/scipy-dev/Library/lib
    name: openblas
    openblas configuration: USE_64BITINT= NO_CBLAS= NO_LAPACK= NO_LAPACKE= DYNAMIC_ARCH=ON
      DYNAMIC_OLDER=OFF NO_AFFINITY=1 USE_OPENMP= HASWELL MAX_THREADS=128
    pc file directory: C:\Users\matth\miniforge3\envs\scipy-dev\Library\lib\pkgconfig
    version: 0.3.27
  lapack:
    detection method: pkgconfig
    found: true
    include directory: C:/Users/matth/miniforge3/envs/scipy-dev/Library/include/openblas
    lib directory: C:/Users/matth/miniforge3/envs/scipy-dev/Library/lib
    name: openblas
    openblas configuration: USE_64BITINT= NO_CBLAS= NO_LAPACK= NO_LAPACKE= DYNAMIC_ARCH=ON
      DYNAMIC_OLDER=OFF NO_AFFINITY=1 USE_OPENMP= HASWELL MAX_THREADS=128
    pc file directory: C:\Users\matth\miniforge3\envs\scipy-dev\Library\lib\pkgconfig
    version: 0.3.27
  pybind11:
    detection method: pkgconfig
    include directory: C:/Users/matth/miniforge3/envs/scipy-dev/Library/include
    name: pybind11
    version: 2.12.0
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.10
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  pythran:
    include directory: ..\..\miniforge3\envs\scipy-dev\Lib\site-packages\pythran
    version: 0.15.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
Python Information:
  path: C:\Users\matth\miniforge3\envs\scipy-dev\python.exe
  version: '3.12'
```

**Comment by nickodell** [comment]
I looked at the implementation, and this appears to be intentional:

https://github.com/scipy/scipy/blob/5383624e90f02a3f1924ee8439a41be919bd8eb5/scipy/special/special/cephes/chdtr.h#L164-L178

To find why this change was made, I went looking through `git blame`, ignoring commits which only move code around. It appears to have been added in commit 867539341e.

Here's the diff from that commit, edited for length:

```diff
commit 867539341eb68e0a0ead13fc6c2170541fd5b41a
Author: Travis Oliphant <oliphant@enthought.com>
Date:   Tue Feb 12 09:39:43 2002 +0000

    Added help for strings which searches scipy tree for object.  Changed stats names in special subdirectory.  Cosmetic changes to cephes *.c files (removed some funny control characters).

[...]

diff --git a/Lib/special/cephes/chdtr.c b/Lib/special/cephes/chdtr.c
index 271217a08..71c61280e 100644
--- a/Lib/special/cephes/chdtr.c
+++ b/Lib/special/cephes/chdtr.c
@@ -157,7 +157,8 @@ double chdtrc(df,x)
 double df, x;
 {
 
-if( (x < 0.0) || (df < 1.0) )
+if (x < 0.0) return 1.0;   /* modified by T. Oliphant */
+if (df < 1.0) 
        {
        mtherr( "chdtrc", DOMAIN );
        return(0.0);

[...]
```

In conclusion, I still don't know why this change was made, but I figured I'd save anyone else the effort of finding this.

**Comment by mdhaber** [comment]
Thanks for tracking that down, but it still doesn't look correct to unconditionally return `1`:
```python3
chdtrc(np.nan, -1)  # 1.0
```

**Comment by fancidev** [comment]
From the C comments in the source code, it seems that the original cephes code returns 0 and sets an error code if x < 0. But _if_ the function is treated as a survival function, a return value of 1 would be more appropriate for negative x. I guess that’s what the previous change was meant to achieve.

The fix is easy — set an error and return nan if df is non-positive. (Btw, the function returns 0 if df is zero; I’m not sure if that’s intended.)

Separately, the [doc](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.chdtrc.html) appears not completely consistent in specifying what the function returns. On the one hand it says the function is a survival function; on the other hand it says the function is an integral (that is not well defined for negative x.)

**Comment by fancidev** [comment]
I'd like to submit a PR for this issue. But I see two possible solutions and I'd like to hear your opinion on which one is preferred: @mdhaber @nickodell @steppi

Solution 1: Fix the immediate issue with `chdtrc` by disallowing (`v <= 0` or `x < 0`). Review the related functions `chdtr`, `chdtri`, `chdtriv` for consistency. (At least, right now `chdtr(1, -1)` returns `nan`, thus violating the desirable identity `chdtr(v, x) + chdtri(v, x) == 1`.) Update the docs to clarify the domain and special values.

Solution 2: Deprecate the four functions `chdtr`, `chdtrc`, `chdtri`, `chdtriv` in favor of the respective gamma functions `gammainc`, `gammaincc`, `gammaincinv`, `gammainccinv` because the `ch*` ones provide little/no value over the `gamma*` ones.

I'm inclined to go for Solution 2 because it saves maintenance burden -- efforts spent to make `ch*` better could better be spent to make `gamma*` better. The drawback is potential backward compatibility break (for user code), which I don't have the visibility to assess.

**Comment by nickodell** [comment]
> Solution 1 maintains backward compatibility, while Solution 2 reduces maintenance burden in the long run. What do you think?

I would argue in favor of Solution 1.

 * Many other projects use these functions. [I searched `chdtrc`](https://github.com/search?q=language%3APython+chdtrc&type=code), and ignoring stuff that wraps cephes or SciPy, or provides a SciPy equivalent API, I count 3 repos which are using this in non-trivial ways on the first page. Assuming 3/20 of the 1.4k results are real users, that would correspond to about 200 users of this function. Removing this function would create churn in those projects.
 * It may be a trivial wrapper around igamc, but that also reduces the maintenance associated with it.

**Comment by steppi** [comment]
I can take a closer look at the actual issue when I have time, but just want to voice that I'm against deprecating and removing`chdtr` and friends. I don't think keeping them around will pose any significant maintenance burden once we've corrected the edge case behavior.

**Comment by fancidev** [comment]
Thanks for the comments @nickodell, @steppi!

So, I proceed to check the edge cases of `chdtr`/`chdtrc` (chi-square distribution functions) and compare them to `gdtr`/`gdtrc` (gamma distribution functions) and `gammainc`/`gammaincc`. The result is summarized below.

```
     v      x    chdtr   chdtrc     gdtr    gdtrc gammainc gammaincc
-------------------------------------------------------------------
 -0.01   -inf      nan   1.0000      nan      nan      nan       nan
 -0.01  -0.01      nan   1.0000      nan      nan      nan       nan
 -0.01   0.00      nan      nan      nan      nan      nan       nan
 -0.01   0.01      nan      nan      nan      nan      nan       nan
 -0.01    inf      nan      nan      nan      nan      nan       nan
-------------------------------------------------------------------
  0.00   -inf      nan   1.0000      nan      nan      nan       nan
  0.00  -0.01      nan   1.0000      nan      nan      nan       nan
  0.00   0.00      nan      nan      nan      nan      nan       nan
  0.00   0.01   1.0000   0.0000   1.0000   0.0000   1.0000    0.0000
  0.00    inf   1.0000   0.0000   1.0000   0.0000   1.0000    0.0000
-------------------------------------------------------------------
  0.01   -inf      nan   1.0000      nan      nan      nan       nan
  0.01  -0.01      nan   1.0000      nan      nan      nan       nan
  0.01   0.00   0.0000   1.0000   0.0000   1.0000   0.0000    1.0000
  0.01   0.01   0.9766   0.0234   0.9766   0.0234   0.9766    0.0234
  0.01    inf   1.0000   0.0000   1.0000   0.0000   1.0000    0.0000
-------------------------------------------------------------------
  1.00   -inf      nan   1.0000      nan      nan      nan       nan
  1.00  -0.01      nan   1.0000      nan      nan      nan       nan
  1.00   0.00   0.0000   1.0000   0.0000   1.0000   0.0000    1.0000
  1.00   0.01   0.0797   0.9203   0.0797   0.9203   0.0797    0.9203
  1.00    inf   1.0000   0.0000   1.0000   0.0000   1.0000    0.0000
-------------------------------------------------------------------
   inf   -inf      nan   1.0000      nan      nan      nan       nan
   inf  -0.01      nan   1.0000      nan      nan      nan       nan
   inf   0.00   0.0000   1.0000   0.0000   1.0000   0.0000    1.0000
   inf   0.01   0.0000   1.0000   0.0000   1.0000   0.0000    1.0000
   inf    inf      nan      nan      nan      nan      nan       nan
-------------------------------------------------------------------
chdtr = chdtr(v,x)
chdtrc = chdtrc(v,x)
gdtr = gdtr(1,v/2,x/2)
gdtrc = gdtrc(1,v/2,x/2)
gammainc = gammainc(v/2,x/2)
gammaincc = gammaincc(v/2,x/2)
```

I make three observations from the above table:

1. `gdtr`/`gdtrc` always agree with `gammainc`/`gammaincc`, so we only have to examine one of them.
2. `chdtrc` "stands out" in how it handles negative `x` (for valid or invalid value of `v`): all the other functions return `nan`, but `chdtrc` returns `1.0`.
3. For `v == 0`, the CDFs return `1.0` if `x > 0` but `nan` if `x == 0`. I'm not sure how `v == 0` is expected to be defined ([this MathOverflow question](https://math.stackexchange.com/questions/4385825/defining-a-chi-square-distribution-with-zero-degrees-of-freedom) suggests `v == 0` corresponds to a point mass at `0`), but at least the current behavior makes the CDFs not right-continuous at zero, which seems not very nice.

So it seems to me that, leaving the behavior of `v == 0` aside, the easiest fix is to align `chdtrc` with the rest in terms of handling `x < 0`, i.e. to return `nan` instead of `1.0`. Do you agree?

**Comment by mdhaber** [comment]
> Do you agree?

I thought that the point of the `_tr`/`_trc` functions was that they would behave like CDFs/CCDFs and return a number for real `x` and valid shape parameter. But I guess that's not the case, so yes, I suppose `chdtrc` should be aligned with the rest w.r.t. `x < 0`.

**Comment by steppi** [comment]
I've been working on rewriting cdflib in C++, with the solver written in a regular structured style instead of the reverse communication style the original authors used to get around the limitations of F77. I guess I'll just make all of the edge case behavior consistent as I go.

**Comment by steppi** [comment]
This was fixed in #22441 as confirmed by the tests here

https://github.com/scipy/scipy/blob/56c8b23935b0e09632bdbbfa6d105945db4f8279/scipy/special/tests/test_basic.py#L4329-L4331

**Comment by steppi** [comment]
This was fixed incidentally in #22441 when we added a strict `NaN` check for `igamc` which `chdtrc` calls

https://github.com/scipy/scipy/blob/56c8b23935b0e09632bdbbfa6d105945db4f8279/scipy/special/xsf/cephes/igam.h#L328-L333

Should we add a test to confirm before closing?

**Comment by mdhaber** [comment]
I agree that this is fixed in main:

```python3
import pandas as pd
import numpy as np
from scipy import special

v = np.asarray([-0.01, 0, 0.01, 1, np.inf])[:, np.newaxis]
x = np.asarray([-np.inf, -0.01, 0, 0.01, np.inf])

chdtr = special.chdtr(v, x)
gdtr = special.gdtr(1, v/2, x/2)
gammainc = special.gammainc(v/2, x/2)
chdtrc = special.chdtrc(v, x)
gdtrc = special.gdtrc(1, v/2, x/2)
gammaincc = special.gammaincc(v/2, x/2)
```

[Code for putting it in a table](https://chatgpt.com/share/67aa8ced-f6e8-8010-bbfa-c7a3209dbba7).

```
       v     x     chdtr      gdtr  gammainc    chdtrc     gdtrc  gammaincc
0  -0.01  -inf       NaN       NaN       NaN       NaN       NaN        NaN
1  -0.01 -0.01       NaN       NaN       NaN       NaN       NaN        NaN
2  -0.01  0.00       NaN       NaN       NaN       NaN       NaN        NaN
3  -0.01  0.01       NaN       NaN       NaN       NaN       NaN        NaN
4  -0.01   inf       NaN       NaN       NaN       NaN       NaN        NaN
5   0.00  -inf       NaN       NaN       NaN       NaN       NaN        NaN
6   0.00 -0.01       NaN       NaN       NaN       NaN       NaN        NaN
7   0.00  0.00       NaN       NaN       NaN       NaN       NaN        NaN
8   0.00  0.01  1.000000  1.000000  1.000000  0.000000  0.000000   0.000000
9   0.00   inf  1.000000  1.000000  1.000000  0.000000  0.000000   0.000000
10  0.01  -inf       NaN       NaN       NaN       NaN       NaN        NaN
11  0.01 -0.01       NaN       NaN       NaN       NaN       NaN        NaN
12  0.01  0.00  0.000000  0.000000  0.000000  1.000000  1.000000   1.000000
13  0.01  0.01  0.976627  0.976627  0.976627  0.023373  0.023373   0.023373
14  0.01   inf  1.000000  1.000000  1.000000  0.000000  0.000000   0.000000
15  1.00  -inf       NaN       NaN       NaN       NaN       NaN        NaN
16  1.00 -0.01       NaN       NaN       NaN       NaN       NaN        NaN
17  1.00  0.00  0.000000  0.000000  0.000000  1.000000  1.000000   1.000000
18  1.00  0.01  0.079656  0.079656  0.079656  0.920344  0.920344   0.920344
19  1.00   inf  1.000000  1.000000  1.000000  0.000000  0.000000   0.000000
20   inf  -inf       NaN       NaN       NaN       NaN       NaN        NaN
21   inf -0.01       NaN       NaN       NaN       NaN       NaN        NaN
22   inf  0.00  0.000000  0.000000  0.000000  1.000000  1.000000   1.000000
23   inf  0.01  0.000000  0.000000  0.000000  1.000000  1.000000   1.000000
24   inf   inf       NaN       NaN       NaN       NaN       NaN        NaN
```
